### PR TITLE
bridge_qinq: Add basic QinQ test

### PR DESCRIPTION
**Pull request by xiywang**
id: 1701557

Signed-off-by: Xiyue Wang <xiywang@redhat.com>

### CODE REVIEW

**Comment by xiywang** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.my.bridge_qinq: PASS (813.30 s)

File: qemu/tests/cfg/bridge_qinq.cfg [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    type = bridge_qinq
+    private_bridge = tmpbr
+    copy_qinq_script = linux_qinq/qinq.sh
+    ping_count = 100
+    net_mask = "24"
+    ip_vm = "192.168.1.2"
+    vlan_id = 10
+    L1tag_iface = "v1v${vlan_id}"
+    L1tag_iface_ip = "192.168.10.10"
+    tcpdump_cmd = "tcpdump -xxvvleni %s > %s &"
```
Comment (PaulYuuu) [inline qemu/tests/cfg/bridge_qinq.cfg]:
```suggestion
    tcpdump_cmd = "tcpdump -xxvvleni %s icmp > %s &"
```

File: qemu/tests/bridge_qinq.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+            logging.error("Could not get file md5sum in guest")
+            return False
+        logging.debug("md5sum: guest(%s), host(%s)", guest_result, host_result)
+        return guest_result == host_result
+
+    login_timeout = int(params.get("login_timeout", "600"))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    guest_qinq_dir = params["guest_qinq_dir"]
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]: Missing this parameter.

Comment (pezhang) [reply qemu/tests/bridge_qinq.py]: Missing it cause:
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.bridge_qinq.q35: SKIP: Mandatory parameter 'guest_qinq_dir' is missing. Check your cfg files for typos/mistakes

File: qemu/tests/bridge_qinq.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    # Set interface v1v10 IP in guest
+    vlan_id = params["vlan_id"]
+    L1tag_iface = params["L1tag_iface"]
+    L1tag_iface_ip = params["L1tag_iface_ip"]
+    session.cmd_output("ifconfig %s %s netmask %s" % (L1tag_iface,
+                                                      L1tag_iface_ip, netmask))
+    output = session.cmd_output("ifconfig %s" % L1tag_iface, timeout=120)
+    logging.info("%s", output)
+
+    # Start tcpdump on L1tag interface and eth0 in guest
```
[the remark this answers is not in the document]
Comment (xiywang) [reply qemu/tests/bridge_qinq.py]: @PaulYuuu Yes I understand that. But I cannot think out a proper name of this interface. What would you suggest?

File: qemu/tests/cfg/bridge_qinq.cfg (hunk 1 above)
Comment (xiywang) [reply qemu/tests/cfg/bridge_qinq.cfg]: If add 'icmp' in the tcpdump cmd, line 288 in py file cannot get any output from tcpdump.

File: qemu/tests/bridge_qinq.py (hunk 2 above)
Comment (xiywang) [reply qemu/tests/bridge_qinq.py]: @pezhang @PaulYuuu Updated. Please help review again. Thanks.

File: qemu/tests/bridge_qinq.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulYuuu) [reply qemu/tests/bridge_qinq.py]: hmm, can be defined as the first nic

File: qemu/tests/cfg/bridge_qinq.cfg (hunk 1 above)
Comment (PaulYuuu) [reply qemu/tests/cfg/bridge_qinq.cfg]: Yes, my fault. `icmp` only only works for L1tag, `ether proto` can grep the ethertype

Comment (PaulYuuu) [reply qemu/tests/cfg/bridge_qinq.cfg]: Yes, my fault. `icmp` only only works for L1tag, `ether proto` can grep the ethertype

File: qemu/tests/bridge_qinq.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        :param vm: guest vm
+
+        """
+        error_context.context("Copy qinq script to guest", logging.info)
+        host_qinq_dir = os.path.join(data_dir.get_deps_dir(),
+                                     params.get("copy_qinq_script"))
+        vm.copy_files_to(host_qinq_dir, guest_qinq_dir)
+
+    def check_tcpdump_result(session, iface_name, ethertype, ethertype2=None,
+                             vlan_tag=None, vlan_tag2=None, enable_logging=True):
```
[the remark this answers is not in the document]
Comment (xiywang) [reply qemu/tests/bridge_qinq.py]: @PaulYuuu Updated. Please help review again.

File: qemu/tests/bridge_qinq.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    12) Do ping test
+    13) Check tcpdump result with vlan tag and ethertype
+    14) SCP file transfer between host and guest
+
+    :param test: KVM test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def copy_qinq_file(session, vm, guest_qinq_dir):
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]: Looks like `session` not be used.

File: qemu/tests/cfg/bridge_qinq.cfg [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    only Linux
+    only bridge
+    type = bridge_qinq
+    private_bridge = tmpbr
+    copy_qinq_script = linux_qinq/qinq.sh
+    guest_qinq_dir = /home/
+    ping_count = 100
+    net_mask = "24"
+    ip_vm = "192.168.1.2"
+    vlan_id = 10
```
Comment (PaulYuuu) [inline qemu/tests/cfg/bridge_qinq.cfg]: Looks like `vlan_id` not be used.

File: qemu/tests/bridge_qinq.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    # Create 802.1ad vlan via bridge in host
+    error_context.context("Create 802.1ad vlan via bridge %s" % brname,
+                          logging.info)
+    advlan_ifname = params["advlan_name"]
+    advlan_ip = params["advlan_ip"]
+    add_advlan_cmd = params["add_advlan_cmd"]
+    process.system_output(add_advlan_cmd)
+    advlan_iface = utils_net.Interface(advlan_ifname)
+    advlan_iface.set_mac(params["advlan_mac"])
+    advlan_iface.set_ip(params["advlan_ip"])
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]:
```suggestion
    advlan_iface.set_ip(advlan_ip)
```
Or remove this one
```python
advlan_ip = params["advlan_ip"]
```

File: qemu/tests/bridge_qinq.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    advlan_iface.set_netmask(int(netmask))
+    advlan_iface.up()
+    output = process.getoutput("ifconfig %s" % advlan_ifname)
+    logging.info("%s" % output)
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]:
```suggestion
    logging.info(output)
```

File: qemu/tests/bridge_qinq.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+    # Stop tcpdump and check result
+    session.cmd_output_safe("pkill tcpdump")
+    check_tcpdump_result(session, L1tag_iface,
+                         "ethertype IPv4 (0x0800)")
+    check_tcpdump_result(session, nic_name,
+                         "ethertype 802.1Q-QinQ (0x88a8)", vlan_tag="vlan 10,")
+
+    # Set IP on L2 tag on the guest interface with vid 20
+    L2vid = params["L2vid"]
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]: This line is not used

File: qemu/tests/bridge_qinq.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    session.sendline(L2tag_tcpdump_cmd)
+    time.sleep(2)
+    session.sendline(first_nic_tcpdump_cmd)
+    time.sleep(5)
+
+    # Create 802.1q vlan via 802.1ad vlan in host
+    error_context.context("Create 802.1q vlan via 802.1ad vlan %s" %
+                          advlan_ifname, logging.info)
+    qvlan_ifname = params["qvlan_name"]
+    qvlan_ip = params["qvlan_ip"]
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]: This line is not used

File: qemu/tests/bridge_qinq.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+    session.cmd_output("ip addr add %s/%s dev %s" % (ip, netmask,
+                                                     nic_name))
+
+    # Create vlans via script qinq.sh
+    output = session.cmd_output("sh %sqinq.sh %s" % (guest_qinq_dir,
+                                                     nic_name), timeout=300)
+    logging.info("%s", output)
+
+    # Set interface v1v10 IP in guest
+    vlan_id = params["vlan_id"]
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]: Looks like `vlan_id` not be used.

**Comment by pezhang** [comment]
(4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.bridge_qinq.q35: PASS (639.49 s)

File: qemu/tests/bridge_qinq.py (hunk 11 above)
Comment (xiywang) [reply qemu/tests/bridge_qinq.py]: @PaulYuuu Updated.

**Comment by xiywang** [comment]
@vivianQizhu Any more comments?

File: qemu/tests/bridge_qinq.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+                          advlan_ifname, logging.info)
+    qvlan_ifname = params["qvlan_name"]
+    add_qvlan_cmd = params["add_qvlan_cmd"]
+    process.system_output(add_qvlan_cmd)
+    qvlan_iface = utils_net.Interface(qvlan_ifname)
+    qvlan_iface.set_ip(params["qvlan_ip"])
+    qvlan_iface.set_netmask(int(netmask))
+    qvlan_iface.up()
+    output = process.getoutput("ifconfig %s" % qvlan_ifname)
+    logging.info("%s" % output)
```
Comment (PaulYuuu) [inline qemu/tests/bridge_qinq.py]: Modify those`logging.info`
```suggestion
    logging.info(output)
```
Other looks good to me.

**Comment by pezhang** [comment]
Looks good to me.

Acked by: Pei Zhang pezhang@redhat.com

**Comment by pezhang** [comment]
Test pass.
 (1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.bridge_qinq.q35: PASS (652.82 s)

**Pull request merged**